fedauth.go

```diff
@@ -0,0 +1,101 @@
+package mssql
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"golang.org/x/crypto/pkcs12"
+)
+
+const (
+	activeDirectoryEndpoint = "https://login.microsoftonline.com/"
+	azureSQLResource        = "https://database.windows.net/"
+	driverClientID          = "7f98cb04-cd1e-40df-9140-3bf7e2cea4db"
+)
+
+func fedAuthGetClientCertificate(clientCertPath, clientCertPassword string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	pkcs, err := ioutil.ReadFile(clientCertPath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to read the AD client certificate from path %s: %v", clientCertPath, err)
+	}
+
+	privateKey, certificate, err := pkcs12.Decode(pkcs, clientCertPassword)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to read the AD client certificate from path %s: %v", clientCertPath, err)
+	}
+
+	rsaPrivateKey, isRsaKey := privateKey.(*rsa.PrivateKey)
+	if !isRsaKey {
+		return nil, nil, fmt.Errorf("AD client certificate at path %s must contain an RSA private key", clientCertPath)
+	}
+
+	return certificate, rsaPrivateKey, nil
+}
+
+func fedAuthGetAccessToken(ctx context.Context, resource, tenantID string, p connectParams, log optionalLogger) (accessToken string, err error) {
```

Comment:
I'd really like to see this function isolated in a sub package at a minimum. Maybe we can still directly reference it, but I would really like it injected somehow.

Reply:
I could try to refactor this so that it uses a package initialization routine to add in the Azure AD support? Would that work? Then a user that wanted Azure AD via connection string parameters would need to use two imports:
```go
import (
	"database/sql"

	_ "github.com/denisenkom/go-mssqldb"
	_ "github.com/denisenkom/go-mssqldb/azure"
)
```
If that's acceptable, I think it would be possible to make the direct dependency opt-in like that.

Reply:
Actual proposal has the import path as github.com/denisenkom/go-mssqldb/azuread, and it is sufficient to just import that package if the generic sql.Open method is used.